### PR TITLE
refactor: Migrate users and remove conversation dependency from meetings

### PR DESCRIPTION
**Pull request by bzang**
Finally getting there. There are a few edge cases that are still concerning, but are incredibly hard to test. A couple of items that still need to be addressed, but could probably be moved to a different PR. For reviewing purposes, going through commit by commit will make more sense.

- [ ] Add snapshot tests for widget components
- [ ] Migrate widget messages and roster complete off of conversations to spaces module
- [ ] Add documentation for widget-meet and widget-message
- [ ] Odd error when a call gets redirected and a locus replace happens

### CODE REVIEW

**Comment by adamweeks** [comment]
Deploy preview for *hopeful-hamilton-4090b9* ready!

Built with commit 5aa5ac1ac21ab60b9cc3640a23aec7fcc68d1644

https://deploy-preview-660--hopeful-hamilton-4090b9.netlify.com

File: packages/node_modules/@ciscospark/widget-recents/src/container.js [hunk 1]
```diff
@@ -242,9 +261,9 @@ export class RecentsWidget extends Component {
           }
           else if (
             s.type === 'group' &&
-            spaces.hasIn(['items', s.id])
+            spaces.has(s.id)
           ) {
-            nextProps.fetchAvatar({space: spaces.getIn(['items', s.id]).toJS()}, sparkInstance);
+            nextProps.fetchAvatar({space: spaces.get(s.id)}, sparkInstance);
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/widget-recents/src/container.js]: ❤️ Records! So much cleaner.

File: packages/node_modules/@ciscospark/redux-module-mercury/src/enhancer.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  setWrappedDisplayName('WithMercury'),
+  connect(
+    (state) => state,
+    (dispatch) => bindActionCreators({
+      connectToMercury
+    }, dispatch)
+  ),
+  lifecycle({
+    componentWillReceiveProps(nextProps) {
+      setupMercury(nextProps);
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-mercury/src/enhancer.js]: Needed this this weekend

File: packages/node_modules/@ciscospark/redux-module-avatar/src/actions.js [hunk 3]
```diff
@@ -40,9 +40,13 @@ function fetchUserAvatar(userId, spark) {
...
     return spark.people.get(userId)
-      .then((person) => dispatch(addAvatar(userId, convertToSmallAvatar(person.avatar))))
+      .then((person) => {
+        const avatar = convertToSmallAvatar(person.avatar);
+        dispatch(addAvatar(userId, avatar));
+        return Promise.resolve(avatar);
+      })
       .catch(() => {
-        dispatch(addAvatar(userId));
+        dispatch(addAvatar({id: userId}));
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-avatar/src/actions.js]: This doesn't look like it would work. `addAvatar` takes two params.

File: packages/node_modules/@ciscospark/widget-message/src/selector.js [hunk 4]
```diff
@@ -50,7 +50,7 @@ const getSpace = createSelector(
   [getConversation, getUsers],
   (conversation, users) => {
     const space = conversation;
-    const currentUser = users.get('currentUser');
+    const currentUser = users.getIn(['byId', users.get('currentUserId')]);
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/widget-message/src/selector.js]: Seems like we could almost use a helper for getting the current user, I've seen this code a bunch.

File: packages/node_modules/@ciscospark/redux-module-teams/src/index.test.js [hunk 5]
```diff
@@ -17,7 +17,16 @@ describe('redux-module-teams actions', () => {
     mockSpark = {
       internal: {
         team: {
-          list: jest.fn(() => Promise.resolve({}))
+          list: jest.fn(() => Promise.resolve([{
+            id: 'teamId',
+            color: '#abcabc',
+            generalConversationId: '10000000-0000-0000-0000-000000000000',
+            displayName: 'test team!',
+            description: 'team that tests',
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/redux-module-teams/src/index.test.js]: nice!

File: packages/node_modules/@ciscospark/widget-message/src/selector.js (hunk 4 above)
Comment (bzang) [reply packages/node_modules/@ciscospark/widget-message/src/selector.js]: Yea. I this is an inconsistency... I actually inject a prop at the top for `currentUser` but here and there I forgot about it.

File: packages/node_modules/@ciscospark/widget-meet/src/enhancers/index.js [hunk 6]
```diff
@@ -1,3 +1,19 @@
...
+import withErrors from './withErrors';
+import withWebRtcSupport from './withWebRtcSupport';
+import withEventHandler from './withEventHandler';
+import withCallHandlers from './withCallHandlers';
+import withToType from './withToType';
+import withMeetDetails from './withMeetDetails';
+import withExternalCall from './withExternalCall';
 
-export default [callMemberships];
+export default [
```
[the remark this answers is not in the document]
Comment (bzang) [reply packages/node_modules/@ciscospark/widget-meet/src/enhancers/index.js]: huh?

File: packages/node_modules/@ciscospark/redux-module-avatar/src/actions.js (hunk 3 above)
Comment (bzang) [reply packages/node_modules/@ciscospark/redux-module-avatar/src/actions.js]: Yea... fixed later. But this catch is actually to just inject an empty placeholder to tell us we've tried to fetch before and to not to try again.

File: packages/node_modules/@ciscospark/redux-module-media/src/enhancer.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    props.listenForIncomingCalls(sparkInstance);
+  }
+}
+
+
+export default compose(
+  setWrappedDisplayName('WithMedia'),
+  connect(
+    (state) => ({
+      sparkInstance: state.spark.get('spark')
```
[the remark this answers is not in the document]
Comment (bzang) [reply packages/node_modules/@ciscospark/redux-module-media/src/enhancer.js]: Good catch

File: packages/node_modules/@ciscospark/widget-meet/src/enhancers/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamweeks) [reply packages/node_modules/@ciscospark/widget-meet/src/enhancers/index.js]: All the handlers. Impressive. Hence, Owen Wilson saying "wow" a lot. Duh. Surprised you couldn't figure that out </s>

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js [hunk 8]
```diff
@@ -102,73 +113,90 @@ export function dismissIncomingCall(callId) {
...
     call.off();
-    return call.hangup().then(() => dispatch(removeCall(call, locusUrl)));
+    return call.hangup().then(() => dispatch(removeCall(call)));
   };
 }
 
 
-function bindEvents(dispatch, call, locusUrl) {
+function bindEvents(dispatch, call) {
+  const updateCall = (callInstance, eventName) => dispatch(updateCallStatus(callInstance, eventName));
```
[the remark this answers is not in the document]
Comment (bzang) [reply packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Yea. I was trying to figure out why a flag was being overwritten and it was impossible to tell which event was doing it. Adding that made it a lot clearer. So there's another philosophy on action naming... next time you get a chance, load up slack and have the redux dev tools open. I'd like to add some of their features, like they fire actions that don't do anything to indicate a thunk was called, and they add explanations to the payloads so it makes things easier to find and debug.
screen shot 2018-02-19 at 4 30 47 pm

**Comment by bzang** [comment]
@adamweeks, i'm inclined to use records for most things that are actually records. There are some issues with records in that they don't seem to `.mergeDeep` correctly and only shallow merge.

I made a couple of changes here that, now I've had a chance to step back and look at everything, aren't that great (e.g. Records at the store root). I'm more inclined to do something like:

``` js
const MainRecord = Record({
  id: '',
  spaceId: ''
  someFlag: false,
  isActive: false
});

const StatusRecord = Record({
});

const initialState = Map({
  byId: Map(),
  byOtherKey: Map(),
  status: StatusRecord()
});
```

Records seem to work best/most consistently when they're completely flat and not nested.

File: packages/node_modules/@ciscospark/redux-module-media/src/actions.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (adamweeks) [reply packages/node_modules/@ciscospark/redux-module-media/src/actions.js]: Yea, that is super nice. I definitely think we can do more (in)actions for debugging purposes.

**Comment by adamweeks** [comment]
@bzang that's good to know. I'm good with initial states not being records and using them really only at the "bottom" level

File: test/journeys/specs/space/global/meet.js [hunk 9]
```diff
@@ -14,21 +14,7 @@ describe('Widget Space', () => {
...
-    'spark:memberships_write',
-    'spark:messages_read',
-    'spark:messages_write',
-    'spark:teams_read',
-    'spark:teams_write',
-    'spark:team_memberships_read',
-    'spark:team_memberships_write',
-    'spark:kms'
-  ].join(' ');
+  process.env.CISCOSPARK_SCOPE = 'spark:all';
```
Comment (adamweeks) [inline test/journeys/specs/space/global/meet.js]: spark:all on this one only?

File: test/journeys/specs/space/global/meet.js [hunk 10]
```diff
@@ -115,6 +101,7 @@ describe('Widget Space', () => {
         },
         spaceId
       };
+      console.warn(options);
```
Comment (adamweeks) [inline test/journeys/specs/space/global/meet.js]: consoles?

File: test/journeys/specs/space/global/meet.js [hunk 11]
```diff
@@ -129,6 +116,7 @@ describe('Widget Space', () => {
...
```
Comment (adamweeks) [inline test/journeys/specs/space/global/meet.js]: consoles?

File: test/journeys/lib/test-helpers/space-widget/meet.js [hunk 12]
```diff
@@ -152,8 +152,8 @@ export function callEventTest(caller, receiver, space = false) {
     assert.containsAllKeys(eventCreated.detail.data, ['actorName', 'roomId']);
     assert.containsAllKeys(receiverEventCreated.detail, ['resource', 'event', 'data']);
     assert.containsAllKeys(receiverEventCreated.detail.data, ['actorName', 'roomId']);
-    assert.equal(eventCreated.detail.data.actorName, space.displayName, 'call created event did not have space name');
-    assert.equal(receiverEventCreated.detail.data.actorName, space.displayName, 'call created event on receiver did not have space name');
+    assert.equal(eventCreated.detail.data.actorName, caller.user.displayName, 'call created event did not have space name');
```
Comment (adamweeks) [inline test/journeys/lib/test-helpers/space-widget/meet.js]: Should the error message on these be `did not have user name`?

File: test/journeys/lib/test-helpers/space-widget/meet.js [hunk 13]
```diff
@@ -152,8 +152,8 @@ export function callEventTest(caller, receiver, space = false) {
...
+    assert.equal(receiverEventCreated.detail.data.actorName, caller.user.displayName, 'call created event on receiver did not have space name');
```
Comment (adamweeks) [inline test/journeys/lib/test-helpers/space-widget/meet.js]: user

File: test/journeys/lib/test-helpers/space-widget/meet.js [hunk 14]
```diff
@@ -166,24 +166,26 @@ export function callEventTest(caller, receiver, space = false) {
...
   const eventConnected = callerEvents.find((event) => event.eventName === 'calls:connected');
-  assert.isDefined(eventConnected, 'has a calls connected event');
-  assert.containsAllKeys(eventConnected.detail, ['resource', 'event', 'actorId', 'data']);
-  assert.containsAllKeys(eventConnected.detail.data, ['actorName', 'roomId']);
+  assert.isDefined(eventConnected, 'has a calls connected event', errorMessage);
+  assert.containsAllKeys(eventConnected.detail, ['resource', 'event', 'actorId', 'data'], errorMessage);
+  assert.containsAllKeys(eventConnected.detail.data, ['actorName', 'roomId'], 'calls:connected', errorMessage);
   if (space) {
-    assert.equal(eventCreated.detail.data.actorName, space.displayName, 'call connected event did not have space name');
+    assert.equal(eventCreated.detail.data.actorName, caller.user.displayName, 'call connected event did not have space name');
```
Comment (adamweeks) [inline test/journeys/lib/test-helpers/space-widget/meet.js]: user

Comment (bzang) [reply test/journeys/lib/test-helpers/space-widget/meet.js]: Yes

File: test/journeys/specs/space/global/meet.js (hunk 11 above)
Comment (bzang) [reply test/journeys/specs/space/global/meet.js]: damn it

**Comment by bzang** [comment]
@adamweeks ok, fixed!
